### PR TITLE
[Whisper] Use Attention Cache

### PR DESCRIPTION
**Pull request by sanchit-gandhi**
# What does this PR do?

Refactors the Whisper model to use the attention cache abstraction proposed in #26681. This is required to have consistency with the `StaticCache` attention class proposed in #27931.

The complexity with the current `Cache` abstraction comes from the fact that Whisper is an encoder-decoder model, meaning each decoder attention layer consists of:
1. A self-attention layer (k/v cache over the previous decoder input ids)
2. A cross-attention layer (k/v cache from the encoder hidden-states)

=> the problematic layer for static generation is the dynamic k/v cache in the self-attention layer. In anticipation of using a static cache for this module, the proposed design uses a separate cache for each layer. We can't build the k/v cache into a single `Cache` abstraction, as the shapes for the self and cross-attention key-values are different (which would break compile).

The design is therefore:

```python
past_key_values: Tuple[Cache] = (past_self_attn_key_values, past_cross_attn_key_values)
```

Where `past_self_attn_key_values` and `past_cross_attn_key_values` are each `Cache` abstractions. This is not the most elegant design, but is compatible with the current `Cache` abstraction. Another option would be to do a refactor of the `Cache` / `DynamicCache` / `StaticCache` for better compatibility with encoder-decoder models.

cc @ArthurZucker @tomaarsen @gante

### CODE REVIEW

File: src/transformers/models/whisper/modeling_whisper.py [hunk 1]
```diff
@@ -864,9 +864,9 @@ def forward(
         hidden_states = self.self_attn_layer_norm(hidden_states)
 
         # Self Attention
-        # decoder uni-directional self-attention cached key/values tuple is at positions 1,2
-        self_attn_past_key_value = past_key_value[:2] if past_key_value is not None else None
-        # add present self-attn cache to positions 1,2 of present_key_value tuple
+        # decoder uni-directional self-attention cached key/values states are at position 0
+        self_attn_past_key_value = past_key_value[0] if past_key_value is not None else None
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:868]: The difficulty comes here from the fact that we're dealing with two sets of past key-values per decoder layer: one from the self-attention, and one from the cross-attention. The current solution uses a separate cache for each.

File: src/transformers/models/whisper/modeling_whisper.py [hunk 2]
```diff
@@ -801,7 +799,7 @@ def forward(
 
 # Copied from transformers.models.mbart.modeling_mbart.MBartDecoderLayer with MBart->Whisper, MBART->WHISPER
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:800]: I'll propagate changes to all other MBart derived modules when we're happy with the design

**Comment by soumendukrg** [comment]
I have a fully functional working solution for static shaped Whisper, which we have extensively tested on librispeech dataset and get same accuracy as original model.

**Comment by sanchit-gandhi** [comment]
Thanks for the context @gante! Is there anything I can do to help with the static cache refactor? Pretty keen to implement a compile-compatible cache for Whisper!

**Pull request closed without merging**